proxy/reverse.go

```diff
@@ -203,3 +210,42 @@ func maybeSetForwardedFor(req *http.Request) {
 	}
 	req.Header.Set("X-Forwarded-For", clientIP)
 }
+
+const proxyLoopHeaderKey = "X-Etcd-Proxy-Hop-Seen-By"
+
+// recordProxy records proxyID and an endpoint in HTTP headers,
+// in order to detect a proxy loop.
+func recordProxy(req *http.Request, proxyID string, loc url.URL) {
+	pv := proxyID + "=" + loc.String()
+	if prior, ok := req.Header[proxyLoopHeaderKey]; ok {
+		pv = strings.Join(prior, ", ") + ", " + pv
+	}
+	req.Header.Set(proxyLoopHeaderKey, pv)
+}
+
+// checkProxyLoop detects a proxy loop by passing around proxyID
+// and endpoints. It returns the repeating endpoint string and
+// truu if it finds the loop.
+func (p *reverseProxy) checkProxyLoop(req *http.Request) (string, bool) {
+	if prior, ok := req.Header[proxyLoopHeaderKey]; ok {
+		for _, v := range strings.Split(prior[0], ", ") {
+			vs := strings.Split(v, "=")
+			if len(vs) == 2 {
+				if p.proxyID == vs[0] {
+					return vs[1], true
+				}
+			}
+		}
+	}
+	return "", false
+}
+
+// banEndpoint bans the URL from reverseProxy's endpoints.
+func (p *reverseProxy) banEndpoint(rp string) {
+	p.director.Lock()
+	p.director.bannedEndpoints[rp] = struct{}{}
+	p.director.Unlock()
+
+	// update endpoints by calling refresh method.
+	p.director.refresh()
```

Comment:
Thanks! This updates `bannedEndpoints` and calls `refresh`, which checks if the endpoint is found in `bannedEndpoints`. And the refresh method excludes the ones that are in `bannedEndpoints`.

To sum, we call refresh to exclude cyclic endpoint.